analytic_engine/src/instance/write.rs

```diff
@@ -150,6 +150,107 @@ impl EncodeContext {
     }
 }
 
+/// Split the write request into multiple batches whose size is determined by
+/// the `max_bytes_per_batch`.
+struct WriteRowGroupSplitter {
+    /// Max bytes per batch. Actually, the size of a batch is not exactly
+    /// ensured less than this `max_bytes_per_batch`, but it is guaranteed that
+    /// the batch contains at most one more row when its size exceeds this
+    /// `max_bytes_per_batch`.
+    max_bytes_per_batch: usize,
+}
+
+enum SplitResult<'a> {
+    Splitted {
+        encoded_batches: Vec<Vec<ByteVec>>,
+        row_group_batches: Vec<RowGroupSlicer<'a>>,
+    },
+    Integrate {
```

Comment:
It seems this item could be implemented using Splitted with len is 1?
```
struct SplitResult {
  encoded_batches: Vec<Vec<ByteVec>>,
  row_group_batches: Vec<RowGroupSlicer<'a>>,
}
```

Reply:
This will lead to extra cost on allocating a vector if no split happens.

Reply:
That's one small point, I'm not strong with this, just want to remove some duplicated code below.